docs/proposals/20190919-machinepool-api.md

```diff
@@ -0,0 +1,361 @@
+  - “...” (populated by user) and found, expect the infrastructure provider is waiting for bootstrap
+    data to be ready.
+  - Found, expect InfrastructureRef to reference an object such as GoogleManagedInstanceGroup,
+    AWSAutoScaleGroup, or AzureVMSS.
+
+##### Provisioning
+
+``` go
+// MachinePoolPhaseProvisioning is the state when the
+// MachinePool infrastructure is being created.
+MachinePoolPhaseProvisioning = MachinePoolPhase("provisioning")
+```
+
+###### Transition Conditions
+
+- MachinePool.Spec.Template.Spec.Bootstrap.ConfigRef -> Status.Ready is true
+- MachinePool.Spec.Template.Spec..Bootstrap.Data is not <nil>
+
+###### Expectations
+
+- MachinePool’s infrastructure to be in the process of being provisioned.
+
+##### Provisioned
+
+``` go
+// MachinePoolPhaseProvisioned is the state when its
+// infrastructure has been created and configured.
+MachinePoolPhaseProvisioned = MachinePoolPhase("provisioned")
+```
+
+###### Transition Conditions
+
+- MachinePool.Spec.Template.Spec.InfrastructureRef -> Status.Ready is true
+- MachinePool.Status.Replicas is synced from MachinePool.Spec.Template.Spec.InfrastructureRef -> Status.Replicas
+
+###### Expectations
+
+- MachinePool’s infrastructure has been created and the compute resources are configured.
+
+##### Running
+
+``` go
+// MachinePoolPhaseRunning is the MachinePool state when it has
+// become a set of Kubernetes Nodes in a Ready state.
+MachinePoolPhaseRunning = MachinePoolPhase("running")
+```
+
+###### Transition Conditions
+
+- Number of Kubernetes Nodes in a Ready state matching MachinePool.Spec.Selector equal MachinePool.Status.Replicas.
+
+###### Expectations
+
+- MachinePool controller should set MachinePool.Status.NodeRefs.
+
+##### Deleting
+
+``` go
+// MachinePoolPhaseDeleting is the MachinePool state when a delete
+// request has been sent to the API Server,
+// but its infrastructure has not yet been fully deleted.
+MachinePoolPhaseDeleting = MachinePoolPhase("deleting")
+```
+
+###### Transition Conditions
+
+- MachinePool.ObjectMeta.DeletionTimestamp is not <nil>
+
+###### Expectations
+
+- Nodes associated with MachinePool should be cordoned and drained.
+- MachinePool’s resources (Bootstrap and InfrastructureRef) should be deleted first using cascading deletion.
+
+##### Failed
+
+``` go
+// MachinePoolPhaseFailed is the MachinePool state when the system
+// might require user intervention.
+MachinePoolPhaseFailed = MachinePoolPhase("failed")
+```
+###### Transition Conditions
+
+- MachinePool.Status.ErrorReason and/or MachinePool.Status.ErrorMessage is populated.
+
+###### Expectations
+
+- User intervention required.
+
+### Controller Collaboration Diagram
+
+![Figure 1](./images/machinepool-api/figure1.png)
+
+### Risks and Mitigations
+
+#### MachinePool type might not cover all potential infrastructure providers
+
+MachinePool is initially designed to encompass commonality across AWS, GCP, and Azure. CAPI adopting
```

Comment:
Is there some list of features these all share that enable their use to back MachinePools? In other words, what are the assumptions MachinePool needs to make for a provider's scale set to be usable behind a MachinePool?

Reply:
I'll add a section for this.

For v1alpha3 I'd consider the common feature set to be:
- Target Capacity for the set/group
- Rolling update parameters

In the future, if we decide we can support autoscale:
- Min/Max machine replicas in a set/group